core/src/main/java/org/apache/iceberg/FilteredManifest.java

```diff
@@ -95,60 +94,38 @@ public FilteredManifest caseSensitive(boolean isCaseSensitive) {
           entry -> entry != null &&
               evaluator.eval(entry.file().partition()) &&
               metricsEvaluator.eval(entry.file()));
 
+      // note that columns itself could have stats projected, that's ok.
+      // We only drop stats if we have forced stats projection.
+      return CloseableIterable.transform(entries, e -> requireStatsProjection ? e.copyWithoutStats() : e);
     } else {
       return reader.entries(projection(fileSchema, columns, caseSensitive));
     }
   }
 
   CloseableIterable<ManifestEntry> liveEntries() {
-    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
-        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
-      Evaluator evaluator = evaluator();
-      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
-
-      return CloseableIterable.filter(reader.entries(projection(fileSchema, columns, caseSensitive)),
-          entry -> entry != null &&
-              entry.status() != Status.DELETED &&
-              evaluator.eval(entry.file().partition()) &&
-              metricsEvaluator.eval(entry.file()));
-
-    } else {
-      return CloseableIterable.filter(reader.entries(projection(fileSchema, columns, caseSensitive)),
-          entry -> entry != null && entry.status() != Status.DELETED);
-    }
+    return CloseableIterable.filter(allEntries(), entry -> entry != null && entry.status() != Status.DELETED);
   }
 
+  /**
+   * @return an Iterator of DataFile. Makes defensive copies of files before returning
+   */
   @Override
   public Iterator<DataFile> iterator() {
-    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue()) ||
-        (partFilter != null && partFilter != Expressions.alwaysTrue())) {
-      Evaluator evaluator = evaluator();
-      InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
-
-      // ensure stats columns are present for metrics evaluation
-      List<String> projectColumns = Lists.newArrayList(columns);
-      projectColumns.addAll(STATS_COLUMNS); // order doesn't matter
-
-      // if no stats columns were projected, drop them
-      boolean dropStats = Sets.intersection(Sets.newHashSet(columns), STATS_COLUMNS).isEmpty();
-
-      return Iterators.transform(
-          Iterators.filter(reader.iterator(partFilter, projection(fileSchema, projectColumns, caseSensitive)),
-              input -> input != null &&
-                  evaluator.eval(input.partition()) &&
-                  metricsEvaluator.eval(input)),
-          dropStats ? DataFile::copyWithoutStats : DataFile::copy);
-
-    } else {
-      return Iterators.transform(
-          reader.iterator(partFilter, projection(fileSchema, columns, caseSensitive)),
-          DataFile::copy);
-    }
+    // If requireStatsProjection is true, we don't create a defensive
+    // copy of manifest entry as it was already done by allEntries()
+    boolean requireStatsProjection = requireStatsProjection();
```

Comment:
Doesn't this check need to be a little different?

If I project just one of the stats columns, then this will be true to pick up the column stats for filtering. But returning `copyWithoutStats()` in that case will drop a column that was projected. Instead, I think we need to check whether the intersection of stats columns and the requested projection is non-empty. If it is non-empty, then copy. If it is empty, copy without stats.

Reply:
> If I project just one of the stats columns, then this will be true to pick up the column stats for filtering

In this case, the `requireStatsProjection()` will return false. Since `Sets.intersection(Sets.newHashSet(columns), STATS_COLUMNS).isEmpty()` will be `false`

Reply:
Okay, I'm thinking of this in two ways: `allStatsProjected(columns)` and `anyStatsProjected(columns)`.

In `allEntries`, we need to check whether we have `allStatsProjcted`. If not, then we need to add them. Here, we need `anyStatsProjected`: if any were requested then we return a `copy` and we can discard stats if not.

Looks like `requireStatsProjection` implements `anyStatsProjected` because only one stats column is required to make `isEmpty` return false. So this method is correct, but the use above is not correct because it will not add stats columns when some, but not all, are missing.

Reply:
Yes. I'll make those changes

Reply:
Thanks @rdblue ! Address your comments, but I named the methods slightly differently.  I do feel the code is slight harder to understand. Do let me know if there's a way to make it simpler, which we can take up in a follow up PR.